packages/fxa-auth-server/lib/routes/subscriptions/paypal.ts

```diff
@@ -135,6 +113,137 @@ export class PayPalHandler extends StripeHandler {
     };
   }
+
+  /**
+   * Updates the billing agreement for a user with a new PayPal token.
+   *
+   * Updates are not allowed if there's:
+   *  - No paypal sourced subscriptions active/past_due
+   *  - An existing billing agreement on file.
+   *
+   * If there's open invoices, attempts to pay them will be made in
+   * the background.
+   */
+  async updatePaypalBillingAgreement(request: AuthRequest) {
+    this.log.begin('subscriptions.createSubscriptionWithPaypal', request);
+    const { uid, email } = await handleAuth(this.db, request.auth, true);
+    await this.customs.check(request, email, 'createSubscriptionWithPaypal');
+
+    let customer = await this.stripeHelper.customer({
+      uid,
+      email,
+    });
+
+    if (!customer) {
+      throw error.unknownCustomer(uid);
+    }
+
+    if (this.stripeHelper.getCustomerPaypalAgreement(customer)) {
+      throw error.billingAgreementExists(customer.id);
+    }
+
+    const paypalSubscription = customer.subscriptions?.data.find(
+      (sub) =>
+        ['active', 'past_due'].includes(sub.status) &&
+        sub.collection_method === 'send_invoice'
+    );
+    if (!paypalSubscription || !customer.currency) {
+      throw error.internalValidationError('updatePaypalBillingAgreement', {
+        customerId: customer.id,
+        message: 'Cannot add PayPal with no paypal subscriptions.',
+        uid,
+      });
+    }
+
+    const { token } = request.payload as Record<string, string>;
+    const { agreementId } = await this.createAndVerifyBillingAgreement({
+      uid,
+      token,
+      currency: customer.currency,
+    });
+
+    await this.stripeHelper.updateCustomerPaypalAgreement(
+      customer,
+      agreementId
+    );
+
+    const now = Date.now() / 1000;
+    const invoices = await this.stripeHelper
+      .fetchOpenInvoices(now, customer.id)
+      .autoPagingToArray({ limit: 10 });
+    if (invoices.length) {
+      for (const invoice of invoices) {
+        this.processInvoiceInBackground(request, customer, invoice);
+      }
+    }
+    await this.customerChanged(request, uid, email);
+
+    this.log.info('subscriptions.updatePaypalBillingAgreement.success', {
+      uid,
+    });
+    return filterCustomer(customer);
+  }
+
+  /**
+   * Processes an invoice in the background with proper sentry reporting
+   * of the originating request that triggered it.
+   */
+  private async processInvoiceInBackground(
+    request: AuthRequest,
+    customer: Stripe.Customer,
+    invoice: Stripe.Invoice
+  ) {
+    try {
+      if (invoice.amount_due === 0) {
+        await this.paypalHelper.processZeroInvoice(invoice);
+      } else {
+        await this.paypalHelper.processInvoice({ customer, invoice });
+      }
+    } catch (err) {
+      reportSentryError(err, request);
+    }
+  }
+
+  /**
+   * Create and verify a billing agreement is funded from the appropriate
+   * country given the currency of the billing agreement.
```

Comment:
given the currency of the customer - and maybe pass in the customer instead? Only a suggestion but this method feels a bit divorced from it's context. Maybe `createAndVerifyBillingAgreementForExistingPaypalCustomer` and then checks for existing subscriptions and currency could happen in here too?

Reply:
This is called from two contexts, in one of them, its being used for the first time where there has never been a subscription, so the customer object will not have a `currency` attribute. In that case it needs the plan currency, in the other context it has the customer currency. That's the main reason it doesn't have any additional logic folded into it, its just setting up the billing agreement and attaching it to the customer while verifying the billing agreement is in the right country for the currency chosen.

Reply:
got it, i missed the ref in createsubscription. no more late night code review.